File: src/providers/openai-cua/rules/check-response-status-incomplete.ts
/**
 * Flags a function that calls responses.create() and treats "no tool calls
 * in the output" as a successful completion, without ever checking
 * response.status === 'incomplete' anywhere in that function — a
 * token-budget truncation can produce the same shape and be silently
 * misreported as a voluntary, successful finish.
 */
import { findResponsesCreateCall } from '../utils.js';

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'A completion check must rule out response.status === "incomplete" before reporting success',
      category: 'reliability',
      rationale:
        'The Responses API can return status "incomplete" with zero visible message output when generation is cut off by the token budget — the same shape ("no computer_call, no message") a purely structural completion check treats as "the model voluntarily finished." Without checking response.status, a token-budget truncation can be silently reported as a successful task completion.',
      docsUrl: 'https://developers.openai.com/api/docs/guides/tools-computer-use',
      recommended: true,
    },
    messages: {
      missingIncompleteCheck:
        'This treats the absence of tool calls as a successful completion, but never checks response.status === "incomplete" — a token-budget truncation can be misreported as success.',
    },
  },
  create(context: any) {
    type FnState = { sawCreateCall: boolean; sawSuccessReturn: boolean; successNode: any; sawStatusCheck: boolean };
    const stack: any[] = [];
    const states = new Map<any, FnState>();

    function ensureState(fn: any): FnState {
      let s = states.get(fn);
      if (!s) {
        s = { sawCreateCall: false, sawSuccessReturn: false, successNode: null, sawStatusCheck: false };
        states.set(fn, s);
      }
      return s;
    }

    function top(): any {
      return stack[stack.length - 1];
    }

    function pushScope(node: any) {
      stack.push(node);
      ensureState(node);
    }

    function popScope() {
      stack.pop();
    }

    function propName(node: any): string | undefined {
      if (!node) return undefined;
      if (node.type === 'Identifier') return node.name;
      if (node.type === 'Literal' && typeof node.value === 'string') return node.value;
      return undefined;
    }

    function isSuccessObjectLiteral(node: any): boolean {
      if (node?.type !== 'ObjectExpression') return false;
      return (node.properties ?? []).some((p: any) => {
        if (p?.type !== 'Property') return false;
        const keyName = propName(p.key);
        return (
          (keyName === 'success' || keyName === 'completed') &&
          p.value?.type === 'Literal' &&
          p.value.value === true
        );
      });
    }

    return {
      Program(node: any) {
        pushScope(node);
      },
      'Program:exit'() {
        for (const state of states.values()) {
          if (state.sawCreateCall && state.sawSuccessReturn && !state.sawStatusCheck) {
            context.report({ node: state.successNode, messageId: 'missingIncompleteCheck' });
          }
        }
      },

      FunctionDeclaration(node: any) {
        pushScope(node);
      },
      'FunctionDeclaration:exit'() {
        popScope();
      },
      FunctionExpression(node: any) {
        pushScope(node);
      },
      'FunctionExpression:exit'() {
        popScope();
      },
      ArrowFunctionExpression(node: any) {
        pushScope(node);
      },
      'ArrowFunctionExpression:exit'() {
        popScope();
      },

      CallExpression(node: any) {
        if (findResponsesCreateCall(node)) {
          const fn = top();
          if (fn) ensureState(fn).sawCreateCall = true;
        }
      },

      ObjectExpression(node: any) {
        if (isSuccessObjectLiteral(node)) {
          const fn = top();
          if (fn) {
            const state = ensureState(fn);
            if (!state.sawSuccessReturn) {
              state.sawSuccessReturn = true;
              state.successNode = node;
            }
          }
        }
      },

      BinaryExpression(node: any) {
        if (node.operator !== '===' && node.operator !== '==') return;
        const sides = [node.left, node.right];
        const statusSide = sides.find((s) => s?.type === 'MemberExpression' && propName(s.property) === 'status');
        const valueSide = sides.find((s) => s !== statusSide);
        if (statusSide && valueSide?.type === 'Literal' && valueSide.value === 'incomplete') {
          const fn = top();
          if (fn) ensureState(fn).sawStatusCheck = true;
        }
      },
    };
  },
};

export const openaiCuaCheckResponseStatusIncompleteRule = rule;

File: src/providers/openai-cua/rules/python/check_response_status_incomplete.py
"""openai-cua-check-response-status-incomplete

Parity with the JS rule: a function that calls responses.create() and treats
the absence of tool calls as a successful completion, without ever checking
response.status == "incomplete" anywhere in that function.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import enclosing_scope_map, loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import is_responses_create_call  # noqa: E402

RULE_KEY = "openai-cua-check-response-status-incomplete"

SUCCESS_KEYS = frozenset({"success", "completed"})


def _is_success_dict(node: ast.AST) -> bool:
    if not isinstance(node, ast.Dict):
        return False
    for k, v in zip(node.keys, node.values):
        if (
            isinstance(k, ast.Constant)
            and k.value in SUCCESS_KEYS
            and isinstance(v, ast.Constant)
            and v.value is True
        ):
            return True
    return False


def _is_status_incomplete_check(node: ast.AST) -> bool:
    if not isinstance(node, ast.Compare) or len(node.ops) != 1 or not isinstance(node.ops[0], ast.Eq):
        return False
    left, right = node.left, node.comparators[0]

    def is_status_attr(n: ast.AST) -> bool:
        return isinstance(n, ast.Attribute) and n.attr == "status"

    def is_incomplete_literal(n: ast.AST) -> bool:
        return isinstance(n, ast.Constant) and n.value == "incomplete"

    return (is_status_attr(left) and is_incomplete_literal(right)) or (
        is_status_attr(right) and is_incomplete_literal(left)
    )


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    scope_of = enclosing_scope_map(tree)
    saw_create_call: set[int] = set()
    success_node_by_scope: dict[int, ast.AST] = {}
    saw_status_check: set[int] = set()

    for node in ast.walk(tree):
        scope = scope_of.get(id(node), tree)
        scope_id = id(scope)

        if isinstance(node, ast.Call) and is_responses_create_call(node):
            saw_create_call.add(scope_id)
        if _is_success_dict(node):
            success_node_by_scope.setdefault(scope_id, node)
        if _is_status_incomplete_check(node):
            saw_status_check.add(scope_id)

    out: list[dict] = []
    for scope_id, node in success_node_by_scope.items():
        if scope_id in saw_create_call and scope_id not in saw_status_check:
            line, col, _, _ = loc(node)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})
    return out
